File: src/main/java/org/apache/prepbuddy/typesystem/BaseDataType.java
package org.apache.prepbuddy.typesystem;

import java.io.Serializable;
import java.util.List;

public enum BaseDataType implements Serializable {

    NUMERIC(DataType.ZIP_CODE_US,
            DataType.MOBILE_NUMBER,
            DataType.IP_ADDRESS,
            DataType.LATITUDE,
            DataType.LONGITUDE,
            DataType.INTEGER,
            DataType.DECIMAL
    ),
    STRING(DataType.CURRENCY,
            DataType.EMAIL,
            DataType.URL,
            DataType.SOCIAL_SECURITY_NUMBER,
            DataType.ZIP_CODE_US,
            DataType.COUNTRY_CODE_2_CHARACTER,
            DataType.COUNTRY_CODE_3_CHARACTER,
            DataType.COUNTRY_NAME,
            DataType.TIMESTAMP
    );

    private final static String PATTERN = "^([+-]?\\d+?\\s?)(\\d*(\\.\\d+)?)+$";
    private final DataType[] subtypes;

    BaseDataType(DataType... subtypes) {
        this.subtypes = subtypes;
    }

    public static BaseDataType getBaseType(List<String> samples) {
        if (matchesWith(PATTERN, samples))
            return NUMERIC;
        return STRING;
    }

    private static boolean matchesWith(String regex, List<String> samples) {
        int counter = 0;
        int threshold = samples.size() / 2;
        for (String string : samples)
            if (string.matches(regex)) counter++;
        return (counter >= threshold);
    }

    public DataType actualType(List<String> sampleData) {
        for (DataType subtype : this.subtypes) if (subtype.isOfType(sampleData)) return subtype;
        return DataType.ALPHANUMERIC_STRING;
    }
}

File: src/main/scala/org/apache/datacommons/prepbuddy/qualityanalyzers/BaseDataType.scala
package org.apache.datacommons.prepbuddy.qualityanalyzers

abstract class BaseDataType extends Serializable {
    def actualType(sampleData: List[String]): DataType

    protected def checkActualType(sampleData: List[String], subtypes: Array[DataType]): DataType = {
        subtypes.find(_.isOfType(sampleData)).getOrElse(ALPHANUMERIC_STRING)
    }
}

object STRING extends BaseDataType {
    val subtypes: Array[DataType] = Array(
        EMPTY,
        CURRENCY,
        EMAIL,
        URL,
        SOCIAL_SECURITY_NUMBER,
        ZIP_CODE_US,
        COUNTRY_CODE_2_CHARACTER,
        COUNTRY_CODE_3_CHARACTER,
        COUNTRY_NAME,
        TIMESTAMP,
        CATEGORICAL_STRING
    )

    override def actualType(sampleData: List[String]): DataType = {
        super.checkActualType(sampleData, subtypes)
    }
}

object NUMERIC extends BaseDataType {
    val subtypes: Array[DataType] = Array(
        ZIP_CODE_US,
        MOBILE_NUMBER,
        CATEGORICAL_INTEGER,
        INTEGER,
        LATITUDE,
        LONGITUDE,
        DECIMAL
    )

    override def actualType(sampleData: List[String]): DataType = {
        super.checkActualType(sampleData, subtypes)
    }
}

object ALPHANUMERIC extends BaseDataType {
    val subtypes: Array[DataType] = Array(
        ZIP_CODE_US,
        MOBILE_NUMBER,
        IP_ADDRESS
    )

    override def actualType(sampleData: List[String]): DataType = {
        super.checkActualType(sampleData, subtypes)
    }
}



